vumi/utils.py

```diff
@@ -119,13 +121,66 @@ def connectionLost(self, reason):
 def http_request_full(url, data=None, headers={}, method='POST',
                       timeout=None, data_limit=None, context_factory=None,
                       agent_class=None, reactor=None):
+    """
+    This is a drop in replacement for the original `http_request_full` method
+    but it has its internals completely replaced by treq. Treq supports SNI
+    and our implementation does not for some reason. Also, we do not want
+    to continue maintaining this because we're favouring treq everywhere
+    anyway.
+
+    """
+    agent_class = agent_class or Agent
+    if reactor is None:
+        # The import replaces the local variable.
+        from twisted.internet import reactor
+    kwargs = {'pool': default_pool(reactor, pool=None, persistent=False)}
+    if context_factory is not None:
+        kwargs['contextFactory'] = context_factory
+    agent = agent_class(reactor, **kwargs)
+    client = HTTPClient(agent)
+
+    def handle_response(response):
+        return SimplishReceiver(response, data_limit).deferred
+
+    d = client.request(method, url, headers=headers, data=data)
+    d.addCallback(handle_response)
+
+    if timeout is not None:
+        cancelling_on_timeout = [False]
```

Comment:
It's a hack to get a mutable value that can be modified elsewhere:
```
>>> foo = [1]; print foo, id(foo)
[1] 4359988792
>>> foo[0] = 2; print foo, id(foo)
[2] 4359988792
```
This lets us pass `cancelling_on_timeout` through a bunch of other code (which can modify the contents) and then check it later to see if it's been changed. Note that this pattern also requires modifications to be of the form `foo[0] = 3` (update the existing list) rather than `foo = [3]` (replace the list with a new one).